File: 1-6/triangle.cpp
# include <bits/stdc++.h>
using namespace std;

int solve(vector<int> a) {
  int n = a.size();
  int ans = 0;
  for (int i=0; i<n; i++) {
    for (int j=i+1; j<n; j++) {
      for (int k=j+1; k<n; k++) {
	int len = a.at(i) + a.at(j) + a.at(k);
	int max_len = max({a.at(i), a.at(j), a.at(k)});
	int rest_len = len - max_len;
	if ( max_len < rest_len ) {
	  ans = max (ans, len);
	}
      }
    }
  }
  return ans;
}

int main() {
  int n1 = 5;
  int n2 = 4;
  vector<int> a1 = {2, 3, 4, 5, 10};
  vector<int> a2 = {4, 5, 10, 20};

  int ans = solve(a2);

  cout << ans << "\n";
  
}

File: 1-6/triangle.ml
let n = 4
let a = [4; 5; 10; 20]

let rec solve ml i j k =
  if i >= n then ml
  else if j >= n then solve ml (i+1) (i+2) (i+3)
  else if k >= n then solve ml i (j+1) (j+2)
  else 
  let x = List.nth a i in
  let y = List.nth a j in
  let z = List.nth a k in
  
  let len = x + y + z in
  let max_len  = max x (max y z) in
  let rest_len = len - max_len in

  if ( max_len < rest_len ) then solve (max ml len) i j (k+1)
  else solve ml i j (k+1)

let ans = solve 0 0 1 2
